laragurol/issue10

laragurol opened the pull request:
get /events/id
- add
1. controller
2. route
3. documentation

*does not include test because of the issues of the env file

laragurol committed:
doc swagger

muhammed-shihebi commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
This file should not be submitted here on GitHub, it should be private on you machine.

muhammed-shihebi commented on src/routes/events.js:
```diff
@@ -0,0 +1,9 @@
+
+const eventsController = require('../controllers/events');
+
+router.get('/:id', eventsController.getOneEvent);
```
You are not validating the id parameter. Someone can pass anything here when they try to use this endpoint.

muhammed-shihebi commented on src/docs/events.yaml:
```diff
@@ -0,0 +1,204 @@
+paths:
```
Add these lines so your endpoint get grouped with other event related endpoint 
```
tags:
  - name: Event
    description: API to manage Events.
```

muhammed-shihebi commented on src/controllers/events.js:
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable consistent-return */
+const eventModel = require('../models/event');
```
Models should be named with a capital letter `EventModel`

muhammed-shihebi commented on src/controllers/events.js:
```diff
@@ -0,0 +1,25 @@
+    try {
+        const event = await eventModel
+            .findById(id)
+            .populate('publisherId')
```
Populating the publisher like this will result in a big user object that contains even the hashedPassword property of the user model. to avoid this, use the populate function like this: 
```
const requiredUserField = [
    'id',
    'firstName',
    'lastName',
    'email',
    'profileImage',
];

event = await EventModel.findById(id)
    .populate("publisherId", requiredUserField.join(" "))
    .populate("confirmedVolunteers", requiredUserField.join(" "))
    .populate("invitedVolunteers", requiredUserField.join(" "));


```

muhammed-shihebi commented on src/docs/events.yaml:
```diff
@@ -0,0 +1,204 @@
+        type: "array"
+        items:
+         type: "object"
+         $ref: "#components/schemas/user"
```
1. **#/components/schemas/user**
2. After you populate the users as I suggested, you have to change this to a ShortUser schema: 
```
ShortUser:
  properties:
    _id:
      type: string
      example: 62e980586aca79a3936917e6
    firstName:
      type: string
      example: John
    lastName:
      type: string
      example: Doe
    email:
      type: string
      example: mail@example.com
    __t:
      type: string
      example: User
    profileImage:
      type: string
      example: https://www.example.com/image.jpg
    fullName:
      type: string
      example: John Doe
    id:
      type: string
      example: 62e980586aca79a3936917e6

```

muhammed-shihebi commented on src/docs/events.yaml:
```diff
@@ -0,0 +1,204 @@
+          description: "succesful operation"
+          schema:
+            $ref: "#/components/schemas/Event"
+        "404":
```
Your error messages also have a schema, something like: 
```
schema: 
  type: object
  properties: 
    message: 
      type: string
      example: Invalid event ID supplied, Event not found

muhammed-shihebi commented on src/docs/events.yaml:
```diff
@@ -0,0 +1,204 @@
+       - "location"
+     properties:
+       publisherId:
+        type: "integer"
```
Since this is populated, replace the type with: 
```
type: "object"
$ref: "#components/schemas/ShortUser"
```
What ShortUser is, is explained in the next comment

muhammed-shihebi commented on src/docs/events.yaml:
```diff
@@ -0,0 +1,204 @@
+        format: "objectId"
+       title:
+        type: "string"
+        description: "Event Title"
```
Be consistent with your documentation. Sometimes you have a description, sometimes not. I suggest that for primitive types, instead of a description, you can provide an example.

muhammed-shihebi commented on src/docs/events.yaml:
```diff
@@ -0,0 +1,204 @@
+          in: "path"
+          description: "ID of the event to return"
+          required: true
+          type: "integer"
```
The id is actually a string, not an integer. Change this wherever you use an Id.

muhammed-shihebi commented on src/docs/events.yaml:
```diff
@@ -0,0 +1,204 @@
+        type: "array"
+        items:
+         type: "object"
+         $ref: "#components/schemas/user"
```
Same thing as confirmedVolunteers.

muhammed-shihebi commented on src/docs/events.yaml:
```diff
@@ -0,0 +1,204 @@
+        items:
+         type: "object"
+         $ref: "#components/schemas/user"
+       address:
```
The address should be defined like this: 
```
address:
  type: object
  properties:
    city:
      type: string
      example: City Name
    country:
      type: string
      example: Country Name
    addressLine:
      type: string
      example: Address Line

```

muhammed-shihebi commented on src/docs/events.yaml:
```diff
@@ -0,0 +1,204 @@
+           type: "string"
+          addressLine:
+           type: "string" 
+       location:
```
Location is an object not an array
```
location:
  type: object
  properties:
    lat:
      type: float
      example: 41.01
    lon:
      type: float
      example: 28.97

```

muhammed-shihebi reviewed (changes requested):
I have spotted some problems in your code and left you some comments. I hope this helps.

laragurol committed:
merge

laragurol committed:
finished get/event/id

muhammed-shihebi commented on src/app.js:
```diff
@@ -34,6 +35,7 @@ app.use(authMiddleware);
 
 app.use('/api/auth', authRouter);
 app.use('/api/event', eventRouter);
+app.use('/api/events', eventRouter);
```
Please use the same endpoint as the other endpoints `/api/evnet`

muhammed-shihebi commented on src/controllers/events.js:
```diff
@@ -180,8 +182,37 @@ async function joinedVolunteers(req, res) {
+            'email',
+            'profileImage',
+        ];
+        if (!mongoose.Types.ObjectId.isValid(id)) {
```
In the validation, you are already checking for the id, so there is no need to do this here again.

muhammed-shihebi commented on src/docs/event.yaml:
```diff
@@ -188,6 +189,50 @@ paths:
     /api/event/:id:
+        get:
+          tags: 
+          - "Events"
```
`- Event`

muhammed-shihebi commented on src/docs/event.yaml:
```diff
@@ -188,6 +189,50 @@ paths:
+              description: "ID of the event to return"
+              required: true
+              type: "string"
+              format: "int64"
```
Delete this, the id is not int64

muhammed-shihebi commented on src/docs/event.yaml:
```diff
@@ -188,6 +189,50 @@ paths:
+              type: "string"
+              format: "int64"
+          responses:
+            "200":
```
The response object is not showing the swagger UI. This response should look as follows in order for it to work: 
```
        "200":
          description: "An event by spesific id"
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Event"
```

muhammed-shihebi commented on src/docs/event.yaml:
```diff
@@ -188,6 +189,50 @@ paths:
+                schema:
+                  type: "object"
+                  properties:
+                    message:
```
`error` instead of message

muhammed-shihebi commented on src/docs/event.yaml:
```diff
@@ -188,6 +189,50 @@ paths:
+              required: true
+              type: "string"
+              format: "int64"
+          responses:
```
Add the validation error response too: 
```
        422:
          description: This status is returned when the request is missing required fields or has an invalid values.
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/ValidationError"
```

muhammed-shihebi commented on src/docs/event.yaml:
```diff
@@ -188,6 +189,50 @@ paths:
+                  properties:
+                    message:
+                     type: "string"
+                     example: "Invalid event ID supplied, Event not found"
```
This message does not match what you have written in your endpoint. It should be `Event not found` instead of `Invalid event ID supplied, Event not found`

muhammed-shihebi commented on src/docs/event.yaml:
```diff
@@ -188,6 +189,50 @@ paths:
+                schema:
+                 type: "object"
+                 properties:
+                  message:
```
This should be `error` instead of message

muhammed-shihebi committed:
Fix problems in events.test.js

This file contains several errors and since we are not doing testing anymore and to save time, I just fixed these errors.

muhammed-shihebi commented on src/routes/__tests__/events.test.js:
```diff
@@ -645,3 +646,83 @@ describe('POST /api/event/:id/volunteers', () => {
     });
 });
+
+describe('Fetching events by id', () => {
```
I did not run these tests since we are not doing testing anymore. I hope they are working correctly though.

laragurol replied:
We ran them and they passed

muhammed-shihebi reviewed (changes requested):
Mostly, your work is done, just fix the mentioned problems and the branch is ready to be merged.  I have edited the event.test.js, so **please pull these changes before doing any changes on your machine locally**.

laragurol committed:
fix get/event/id

laragurol committed:
update

laragurol committed:
GET /events/:id
Fixes #10

laragurol committed:
x

laragurol committed:
final version get/event/id with fixes

laragurol committed:
Merge branch 'laragurol/issue10' of https://github.com/RCDD-202203-TUR-BEW/backend-capstone-turkey-pebble-work into laragurol/issue10

laragurol committed:
final get/event/id fixed

muhammed-shihebi committed:
minor fixes

muhammed-shihebi reviewed (approved):
Nice work, ready to be merged.

halitbatur commented:
@laragurol lets get this merged by tonight

laragurol committed:
Merge branch 'dev' into laragurol/issue10
